server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -170,7 +170,17 @@ protected TransportReplicationAction(Settings settings, String actionName, Trans
     @Override
     protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
         assert request.shardId() != null : "request shardId must be set";
-        runReroutePhase(task, request, listener, true);
+        request.incRef();
+        Releasable releaseBytes = Releasables.releaseOnce(request::decRef);
+        boolean success = false;
+        try {
+            runReroutePhase(task, request, ActionListener.runAfter(listener, releaseBytes::close), true);
+            success = true;
+        } finally {
+            if (success == false) {
+                releaseBytes.close();
```

Comment:
`runReroutePhase` already handles all exceptions. I think we should at least add `assert false` here to signal that it is unexpected, but could also see us going without the finally block here.

Reply:
Right, lets just drop the finally block here. We have assertions on the `onFailure` not throwing again now anyway so there's no point in asserting here :)